File: Sources/Domain/ProxyClassifier.swift
struct ProxyCatalog: Equatable, Sendable {
  private let typesByName: [String: String]

  init(typesByName: [String: String]) {
    self.typesByName = typesByName
  }

  func type(for name: String) -> String? {
    typesByName[name]
  }
}

struct ProxyClassification: Equatable, Sendable {
  let category: TrafficCategory
  let unknownReason: UnknownTrafficReason?
}

typealias ProxyTypeResolver =
  @Sendable (
    _ rawType: String,
    _ nativeClassification: ProxyClassification
  ) -> ProxyClassification

enum UnknownTrafficReason: String, Equatable, Sendable {
  case emptyChain
  case missingCatalogEntry
  case ambiguousProxyType
}

struct ProxyClassifier: Sendable {
  private static let concreteProxyTypes: Set<String> = [
    "anytls",
    "http",
    "hysteria",
    "hysteria2",
    "shadowsocks",
    "shadowsocksr",
    "snell",
    "socks5",
    "ssh",
    "trojan",
    "tuic",
    "vless",
    "vmess",
    "wireguard",
  ]

  private let catalog: ProxyCatalog
  private let resolveProxyType: ProxyTypeResolver

  init(
    catalog: ProxyCatalog,
    resolveProxyType: @escaping ProxyTypeResolver = { _, nativeClassification in
      nativeClassification
    }
  ) {
    self.catalog = catalog
    self.resolveProxyType = resolveProxyType
  }

  func classify(chains: [String]) -> ProxyClassification {
    guard let leaf = chains.first, !leaf.isEmpty else {
      return ProxyClassification(category: .unknown, unknownReason: .emptyChain)
    }

    guard let rawType = catalog.type(for: leaf) else {
      return ProxyClassification(category: .unknown, unknownReason: .missingCatalogEntry)
    }

    let nativeClassification = Self.classifyNatively(rawType: rawType)
    return resolveProxyType(rawType, nativeClassification)
  }

  private static func classifyNatively(rawType: String) -> ProxyClassification {
    let normalizedType =
      rawType
      .lowercased()
      .filter { $0.isLetter || $0.isNumber }

    switch normalizedType {
    case "direct":
      return ProxyClassification(category: .direct, unknownReason: nil)
    case "reject", "rejectdrop":
      return ProxyClassification(category: .reject, unknownReason: nil)
    case let value where Self.concreteProxyTypes.contains(value):
      return ProxyClassification(category: .proxy, unknownReason: nil)
    default:
      return ProxyClassification(category: .unknown, unknownReason: .ambiguousProxyType)
    }
  }
}

File: platform/windows/MihomoMeter.Windows.Core/Domain/ProxyClassifier.cs
namespace MihomoMeter.Windows.Core.Domain;

public sealed class ProxyCatalog
{
    private readonly IReadOnlyDictionary<string, string> _typesByName;

    public ProxyCatalog(IReadOnlyDictionary<string, string> typesByName)
    {
        _typesByName = new Dictionary<string, string>(typesByName, StringComparer.Ordinal);
    }

    public string? TypeFor(string name)
    {
        return _typesByName.TryGetValue(name, out var type) ? type : null;
    }
}

public sealed record ProxyClassification(
    TrafficCategory Category,
    UnknownTrafficReason? UnknownReason = null);

public enum UnknownTrafficReason
{
    EmptyChain,
    MissingCatalogEntry,
    AmbiguousProxyType,
}

public delegate ProxyClassification LazyProxyTypeResolver(
    string rawType,
    Func<ProxyClassification> nativeFallback);

public sealed class ProxyClassifier
{
    private static readonly HashSet<string> ConcreteProxyTypes = new(StringComparer.Ordinal)
    {
        "anytls",
        "http",
        "hysteria",
        "hysteria2",
        "shadowsocks",
        "shadowsocksr",
        "snell",
        "socks5",
        "ssh",
        "trojan",
        "tuic",
        "vless",
        "vmess",
        "wireguard",
    };

    private readonly ProxyCatalog _catalog;
    private readonly Func<string, ProxyClassification, ProxyClassification>? _resolveProxyType;
    private readonly LazyProxyTypeResolver? _resolveProxyTypeLazily;

    public ProxyClassifier(
        ProxyCatalog catalog,
        Func<string, ProxyClassification, ProxyClassification>? resolveProxyType = null)
    {
        _catalog = catalog;
        _resolveProxyType = resolveProxyType ?? ((_, nativeClassification) =>
            nativeClassification);
    }

    public ProxyClassifier(
        ProxyCatalog catalog,
        LazyProxyTypeResolver resolveProxyTypeLazily)
    {
        _catalog = catalog;
        _resolveProxyTypeLazily = resolveProxyTypeLazily;
    }

    public ProxyClassification Classify(IReadOnlyList<string> chains)
    {
        if (chains.Count == 0 || chains[0].Length == 0)
        {
            return new ProxyClassification(TrafficCategory.Unknown, UnknownTrafficReason.EmptyChain);
        }

        var rawType = _catalog.TypeFor(chains[0]);
        if (rawType is null)
        {
            return new ProxyClassification(
                TrafficCategory.Unknown,
                UnknownTrafficReason.MissingCatalogEntry);
        }

        if (_resolveProxyTypeLazily is not null)
        {
            return _resolveProxyTypeLazily(rawType, () => ClassifyNatively(rawType));
        }

        var nativeClassification = ClassifyNatively(rawType);
        return _resolveProxyType!(rawType, nativeClassification);
    }

    private static ProxyClassification ClassifyNatively(string rawType)
    {
        var normalizedType = string.Concat(rawType
            .ToLowerInvariant()
            .Where(character => char.IsLetter(character) || char.IsNumber(character)));

        return normalizedType switch
        {
            "direct" => new ProxyClassification(TrafficCategory.Direct),
            "reject" or "rejectdrop" => new ProxyClassification(TrafficCategory.Reject),
            _ when ConcreteProxyTypes.Contains(normalizedType) =>
                new ProxyClassification(TrafficCategory.Proxy),
            _ => new ProxyClassification(
                TrafficCategory.Unknown,
                UnknownTrafficReason.AmbiguousProxyType),
        };
    }
}
